File: src/commands/pull.ts
/**
 * Pull command - Fetch fresh context from Linear/Jira
 *
 * Downloads the latest issue data from the configured backend and
 * writes it to the local context files.
 */

import chalk from 'chalk';
import ora from 'ora';
import { readConfig } from '../lib/config.js';
import {
  generateContext,
  getFullContextPath,
  resolveTaskId,
  writeFullContextFile,
} from '../lib/context-generator.js';
import { resolvePaths } from '../lib/paths.js';
import type { Backend } from '../types.js';
import { BACKEND_ID_PATTERNS } from '../types.js';

export interface PullOptions {
  backend?: Backend;
}

/**
 * Pull fresh context from Linear/Jira
 *
 * @param taskId - Optional task ID (falls back to current task)
 * @param options - Command options
 */
export async function pull(taskId: string | undefined, options: PullOptions): Promise<void> {
  // Resolve task ID
  const resolvedId = resolveTaskId(taskId);

  if (!resolvedId) {
    console.error(chalk.red('Error: No task ID provided and no current task set'));
    console.error(chalk.gray('Usage: mobius pull <task-id>'));
    console.error(chalk.gray('Or set a current task: mobius set-id <task-id>'));
    process.exit(1);
  }

  // Resolve backend from options or config
  const paths = resolvePaths();
  const config = readConfig(paths.configPath);
  const backend = options.backend ?? config.backend;

  // Validate task ID format
  const pattern = BACKEND_ID_PATTERNS[backend];
  if (!pattern.test(resolvedId)) {
    console.error(chalk.red(`Error: Invalid task ID format for ${backend}: ${resolvedId}`));
    console.error(chalk.gray('Expected format: PREFIX-NUMBER (e.g., MOB-123)'));
    process.exit(1);
  }

  // Fetch context
  const spinner = ora({
    text: `Fetching context for ${chalk.cyan(resolvedId)} from ${backend}...`,
    color: 'blue',
  }).start();

  try {
    const context = await generateContext(resolvedId, {
      forceRefresh: true,
    });

    if (!context) {
      spinner.fail(`Failed to fetch context for ${resolvedId}`);
      console.error(chalk.red(`Could not find issue ${resolvedId} in ${backend}`));
      process.exit(1);
    }

    // Write the full context file
    writeFullContextFile(resolvedId, context);

    spinner.succeed(`Context fetched for ${chalk.cyan(resolvedId)}`);

    // Display summary
    console.log('');
    console.log(chalk.bold('Summary:'));
    console.log(`  Parent:     ${chalk.cyan(context.parent.identifier)} - ${context.parent.title}`);
    console.log(`  Status:     ${context.parent.status}`);
    console.log(`  Sub-tasks:  ${context.subTasks.length}`);

    // Status breakdown
    if (context.subTasks.length > 0) {
      const statusCounts = new Map<string, number>();
      for (const task of context.subTasks) {
        const count = statusCounts.get(task.status) || 0;
        statusCounts.set(task.status, count + 1);
      }

      console.log('');
      console.log(chalk.bold('Status breakdown:'));
      for (const [status, count] of statusCounts) {
        const statusColor = getStatusColor(status);
        console.log(`  ${statusColor(status)}: ${count}`);
      }
    }

    console.log('');
    console.log(chalk.gray(`Context written to: ${getFullContextPath(resolvedId)}`));
  } catch (error) {
    spinner.fail(`Failed to fetch context for ${resolvedId}`);
    const message = error instanceof Error ? error.message : String(error);
    console.error(chalk.red(`Error: ${message}`));
    process.exit(1);
  }
}

/**
 * Get chalk color function for a status
 */
function getStatusColor(status: string): typeof chalk {
  switch (status.toLowerCase()) {
    case 'done':
    case 'completed':
      return chalk.green;
    case 'in progress':
    case 'in_progress':
      return chalk.blue;
    case 'blocked':
      return chalk.red;
    case 'ready':
      return chalk.cyan;
    default:
      return chalk.gray;
  }
}

File: rust/mobius/src/commands/pull.rs
//! Pull command - Fetch fresh context from Linear/Jira

use colored::Colorize;

use crate::config::loader::read_config;
use crate::config::paths::resolve_paths;
use crate::context::{
    generate_context, get_full_context_path, resolve_task_id, write_full_context_file,
};
use crate::types::enums::Backend;

pub fn run(task_id: Option<&str>, backend_override: Option<&str>) -> anyhow::Result<()> {
    // Resolve task ID
    let resolved_id = resolve_task_id(task_id);

    if resolved_id.is_none() {
        eprintln!(
            "{}",
            "Error: No task ID provided and no current task set".red()
        );
        eprintln!("{}", "Usage: mobius pull <task-id>".dimmed());
        eprintln!(
            "{}",
            "Or set a current task: mobius set-id <task-id>".dimmed()
        );
        std::process::exit(1);
    }

    let resolved_id = resolved_id.unwrap();

    // Resolve backend
    let paths = resolve_paths();
    let config = read_config(&paths.config_path).unwrap_or_default();
    let backend: Backend = if let Some(b) = backend_override {
        b.parse().unwrap_or(config.backend)
    } else {
        config.backend
    };

    // Validate task ID format
    if !validate_task_id(&resolved_id, &backend) {
        eprintln!(
            "{}",
            format!(
                "Error: Invalid task ID format for {}: {}",
                backend, resolved_id
            )
            .red()
        );
        eprintln!(
            "{}",
            "Expected format: PREFIX-NUMBER (e.g., MOB-123)".dimmed()
        );
        std::process::exit(1);
    }

    println!(
        "Fetching context for {} from {}...",
        resolved_id.cyan(),
        backend
    );

    // Generate context
    match generate_context(&resolved_id, None, false) {
        Ok(Some(context)) => {
            // Write full context file
            write_full_context_file(&resolved_id, &context)?;

            println!("{} Context fetched for {}", "✓".green(), resolved_id.cyan());

            // Display summary
            println!();
            println!("{}", "Summary:".bold());
            println!(
                "  Parent:     {} - {}",
                context.parent.identifier.cyan(),
                context.parent.title
            );
            println!("  Status:     {}", context.parent.status);
            println!("  Sub-tasks:  {}", context.sub_tasks.len());

            // Status breakdown
            if !context.sub_tasks.is_empty() {
                let mut status_counts: std::collections::HashMap<String, usize> =
                    std::collections::HashMap::new();
                for task in &context.sub_tasks {
                    *status_counts.entry(task.status.clone()).or_insert(0) += 1;
                }

                println!();
                println!("{}", "Status breakdown:".bold());
                for (status, count) in &status_counts {
                    let colored_status = match status.to_lowercase().as_str() {
                        "done" | "completed" => status.green().to_string(),
                        "in progress" | "in_progress" => status.blue().to_string(),
                        "blocked" => status.red().to_string(),
                        "ready" => status.cyan().to_string(),
                        _ => status.dimmed().to_string(),
                    };
                    println!("  {}: {}", colored_status, count);
                }
            }

            println!();
            println!(
                "{}",
                format!(
                    "Context written to: {}",
                    get_full_context_path(&resolved_id).display()
                )
                .dimmed()
            );
        }
        Ok(None) => {
            eprintln!("{}", format!("No context found for {}", resolved_id).red());
            std::process::exit(1);
        }
        Err(e) => {
            eprintln!(
                "{}",
                format!("Failed to fetch context for {}", resolved_id).red()
            );
            eprintln!("{}", format!("Error: {}", e).red());
            std::process::exit(1);
        }
    }

    Ok(())
}

fn validate_task_id(task_id: &str, backend: &Backend) -> bool {
    let pattern = match backend {
        Backend::Linear => regex::Regex::new(r"^[A-Z]+-\d+$").unwrap(),
        Backend::Jira => regex::Regex::new(r"^[A-Z]+-\d+$").unwrap(),
        Backend::Local => regex::Regex::new(r"^(LOC-\d+|task-\d+)$").unwrap(),
    };
    pattern.is_match(task_id)
}
